[WIP] Add ability to create module from C#

slide opened the pull request:
### What does this implement/fix? Explain your changes.

This allows doing something like this, which I haven't found any other way to do:

```csharp
var mod = PythonEngine.AddModule("mymod");
var myobj = new object();
mod.SetAttr("myprop".ToPython(), myobj.ToPython());
```

```python
import mymod
print(mymod.myprop)
```

This is very useful for providing objects to an embedded Python engine.

### Does this close any currently open issues?

I could not find any specific issues that would be related.

### Any other comments?

This is an initial PR to get feedback if this would be allowed, I have a build of the library locally with this change and use this in my code, but would like to have it upstream if possible. I will fill out the remaining details and add tests if folks think this would be ok.

### Checklist

Check all those that are applicable and complete.

-   [ ] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [ ] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

lostmsu commented:
We now have https://github.com/pythonnet/pythonnet/blob/master/src/runtime/pymodule.cs
I think it would make sense to give it a static factory method `Create` instead of placing code into `PythonEngine`.

Also, you should probably use https://github.com/pythonnet/pythonnet/blob/16f04e9c0281cd1e4c266399b35ee89069eb36fe/src/runtime/runtime.cs#L1895

slide commented:
Cool, I didn't know about pymodule.cs, I'll definitely migrate there and use the PyModule_New

lostmsu commented on src/embed_tests/TestPyModule.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace Python.EmbeddingTest
+{
+    class TestPyModule
```
I think test class must be `public`.

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -37,5 +40,40 @@ public static PyModule FromString(string name, string code)
             return new PyModule(ref m);
         }
+
+        public static PyModule Create(string name, string filename = "none")
```
This method violates single responsibility principle.

First, `Create` should never return an existing module. Instead, the caller should be able to check if module already exists by name by calling a separate function.

Second, I an uncertain about setting up `__builtins__` by default. It would be great to know why Python does not do it before discarding that potentially good rationale. Users can always manually do `newModule.Dict.Set("__builtins__", GetBuiltIns())`. If this is a common scenario in your product, just make a factory method combining the two.

Third, I don't think adding new module to `sys.modules` right away is a good idea. Especially, overwriting the existing value. This might cause hard to debug surprise issues.

slide replied:
> Second, I am uncertain about setting up `__builtins__` by default. It would be great to know why Python does not do it before discarding that potentially good rationale. Users can always manually do `newModule.Dict.Set("__builtins__", GetBuiltIns())`. If this is a common scenario in your product, just make a factory method combining the two.
> 

I didn't see a public method that would allow access to the __builtins__. There is an internal method, on `Runtime`, but not a public method that could be used outside.

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -37,5 +40,40 @@ public static PyModule FromString(string name, string code)
+            PythonException.ThrowIfIsNull(__builtins__);
+            int rc = Runtime.PyDict_SetItemString(globals, "__builtins__", __builtins__);
+            PythonException.ThrowIfIsNotZero(rc);
+            rc = Runtime.PyDict_SetItemString(globals, "__file__", new BorrowedReference(filename.ToPython().Handle));
```
The default value for `filename` should be `null` and when it is null `__file__` should simply not be set.

LimpingNinja commented:
This is a good add feature, is anyone working on it? I am migrating from C where I create a handful of accessible modules from C using Py_InitModule3() and passing in the methods (C function pointers) and vars. Is there an alternative method here or is this something that would pend this change being actioned?

slide commented:
I need to get back to it and make the changes that were requested. I just have a local compiled version with my changes that I have been using, so I haven't had a huge push to get me back to fixing this up. I'll try and work on it soon.

slide commented:
@lostmsu I think I addressed the feedback you provided, please let me know if there are more changes you would like me to make. If not, I will make this a normal PR and remove the WIP.

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -29,6 +32,13 @@ public PyModule Reload()
             return new PyModule(ref op);
         }
 
+        public void AddToSysModules()
+        {
+            BorrowedReference modules = Runtime.PyImport_GetModuleDict();
+            int rc = Runtime.PyDict_SetItemString(modules, this.Name, this.Reference);
+            PythonException.ThrowIfIsNotZero(rc);
+        }
+
```
Instead of adding potentially even more methods, that have to do with `sys.modules` in `PyModule`, expose a static member `PyModule.SysModules`, and let users do checking and adding on their own.

It should replace `AddToSysModules` and `IsInSysModules`.

slide replied:
BorrowedReference is an internal struct, what is the best thing to return then since this would be a public API?

lostmsu replied:
For `SysModules`? `PyDict` - it is a dictionary.

slide replied:
Sorry, missed the PyDict in your original comment.

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -37,5 +47,39 @@ public static PyModule FromString(string name, string code)
+
+            if (filename != null)
+            {
+                BorrowedReference globals = Runtime.PyModule_GetDict(new BorrowedReference(op.DangerousGetAddress()));
```
should be just `PyModule_GetDict(op)`

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -37,5 +47,39 @@ public static PyModule FromString(string name, string code)
+            {
+                BorrowedReference globals = Runtime.PyModule_GetDict(new BorrowedReference(op.DangerousGetAddress()));
+                PythonException.ThrowIfIsNull(globals);
+                int rc = Runtime.PyDict_SetItemString(globals, "__file__", new BorrowedReference(filename.ToPython().Handle));
```
should be just `filename.ToPython().Reference`

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -37,5 +47,39 @@ public static PyModule FromString(string name, string code)
+            return new PyModule(ref op);
+        }
+
+        public void AddBuiltins()
```
Can you split this into `void SetBuiltins(PyModule builtins)` and `PyModule.Builtins` (or maybe `Runtime.Builtins`)

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -37,5 +47,39 @@ public static PyModule FromString(string name, string code)
+
+        public static PyModule Create(string name, string filename=null)
+        {
+            NewReference op = Runtime.PyModule_New(name);
```
Add a null check for the `name` parameter.

lostmsu commented on src/runtime/pymodule.cs:
```diff
@@ -37,5 +47,39 @@ public static PyModule FromString(string name, string code)
+                int rc = Runtime.PyDict_SetItemString(globals, "__file__", new BorrowedReference(filename.ToPython().Handle));
+                PythonException.ThrowIfIsNotZero(rc);
+            }
+
+            return new PyModule(ref op);
+        }
```
I suggest to make `Create` method private, and add a constructor instead. E.g.
`public PyModule(string name, string filename = null): this(Create(name, filename)) { }`

slide commented:
Feedback hopefully addressed. Thanks for being patient. I still don't know the codebase very well.
